### DISCUSSION TITLE
how to pass paired reads to another process in another script

### DISCUSSION THREAD

**Discussion by complexgenome** [nextflow-io/nextflow#4017] [Q&A] [answered]
Hi, 
I am new to nextflow and this is difficult for me to understand for simple steps such as passing file pairs to another process in another script.
I have two scripts: 1) get paired reads 2) run fastp

**one.nf**

```groovy
params.outdir = 'results'

params.reads1 = "/Users/name/Downloads/tiny/normal/*_R1_xxx.fastq.gz"
params.reads2 = "/Users/name/Downloads/tiny/normal/*_R2_xxx.fastq.gz"

println "reads: $params.reads1"
println "reads: $params.reads2"

Channel.fromPath(params.reads1,checkIfExists: true).view()
Channel.fromPath(params.reads2,checkIfExists: true).view()

include { fastp } from '/Users/name/Documents/name/nextflow_scripts/fastp.nf'

fqpairs_ch=channel.fromFilePairs('/Users/sariys01/Downloads/tiny/normal/*_R{1,2}_xxx.fastq.gz')

process check {

	output: 
	tuple val(fqpairs_ch)

	""" 
	echo "hello from command line \n"
	"""
}

workflow {

	check().view()
	fastp(fqpairs_ch).view()
}
```

**fastp.nf**

```groovy
params.outdir = "./results"

process fastp() {

tag "$sample_id"

input:
        tuple val(sample_id), file(x), file(x1) from fqpairs_ch

output:
println ("$sample_id")

	script:

	""" 
mkdir fastp_trimmed

        fastp \
        --correction \
	--cut_tail \
	--disable_trim_poly_g \
	--length_required 50 \
	--qualified_quality_phred 20 \
	--thread 12 \
	--trim_poly_x \
	--unqualified_percent_limit 20 \
        -i ${x[0]} -I ${x[1]} \
        -o fastp_trimmed/joinedfiles.dat_${x[0]} -O fastp_trimmed/joinedfiles.dat_${x[1]} 
	"""

}
```

Error:

> 
> ERROR ~ No such variable: fqpairs_ch
> 
>  -- Check script 'fastp.nf' at line: 13 or see '.nextflow.log' file for more details
> 
> 

How do I capture the paired reads in fastp process defined in another script?

**Comment by complexgenome** [comment]
@bentsherman Can you please assist? I am new to nextflow and is complicated to understand.

**Comment by bentsherman** [comment]
Hi @sariya , you are mixing DSL1 and DSL2 syntax. In your `fastp` process, the `from fqpairs_ch` needs to be removed. Also, as a best practice you should put all of your channel logic in a workflow block. So the `fqpairs_ch` should be defined in your workflow block, and then you should pass it into the `check` process as an argument. Otherwise that part will also fail because `check` expects one argument.

**Reply by complexgenome** [reply]
@bentsherman Thank you. I'm struggling with nextflow. 

How do I know what's DSL1 syntax and/or otherwise?

I'm terrible sorry as I'm unable to follow you. I've been trying to send argument to `fastp` but I fail repeatedly. 

> So the `fqpairs_ch` should be defined in your workflow block

What is the workflow block?

> and then you should pass it into the `check` process as an argument

I have not been able to do so. Do I decalre fastp as fastp or fastp() in another script?

If I put fastq_ch in workflow block, I get java error

```

workflow {
	check().view()
fqpairs_ch=channel.fromFilePairs('/Users/sariys01/Downloads/tiny/normal/*_R{1,2}_xxx.fastq.gz')
	fastp(fqpairs_ch).view()
}
```

Error: attached.
<img width="1122" alt="Screen Shot 2023-06-13 at 4 30 36 PM" src="https://github.com/nextflow-io/nextflow/assets/5683969/dce662c5-4d98-4850-9235-e7a22ebc9493">

**Reply by bentsherman** [reply, marked as answer]
I gave another look and I am seeing multiple errors, so I will just correct the code and show you what I changed:

- don't need parentheses after `fastp` process name
- reference file pairs as `x0` and `x1` instead of `x[0]` and `x[1]`
- remove `println` from process output, emit your output files here instead

```groovy
process fastp {

    tag "$sample_id"

    input:
    tuple val(sample_id), file(x0), file(x1)

    output:
    val(sample_id)

    script:
    """
    mkdir fastp_trimmed

    fastp \
        --correction \
        --cut_tail \
        --disable_trim_poly_g \
        --length_required 50 \
        --qualified_quality_phred 20 \
        --thread 12 \
        --trim_poly_x \
        --unqualified_percent_limit 20 \
        -i ${x0} -I ${x1} \
        -o fastp_trimmed/joinedfiles.dat_${x0} -O fastp_trimmed/joinedfiles.dat_${x1}
    """
}

workflow {
    fqpairs_ch = channel.fromFilePairs('/Users/sariys01/Downloads/tiny/normal/*_R{1,2}_xxx.fastq.gz')
    fastp(fqpairs_ch).view()
}
```

Let's get this part working first, then it should be easy for you to move your process into a separate module. Try this snippet with the rest of your code and let me know if it works.

**Comment by complexgenome** [comment]
@bentsherman  Thanks for editing and reviewing code. It works, however there's one concern - it doesn't make directory as indicated at the working directory level. It errors out due to mismatch FASTQ paired end reads; however, not creating directory is mysterious. 

[output_log.txt](https://github.com/nextflow-io/nextflow/files/11747559/output_log.txt)
```

ls ./*/*
ls work/22/697720257efeeacd51d21a4f84f2d0/fastp_trimmed
```

Next, how do I keep process in a separate file and send parameters?

Thank you again.

**Reply by bentsherman** [reply]
@sariya , I recommend that you refer to our training resources such as https://training.nextflow.io/ , so that you can acquire a proper understanding of how to use Nextflow. That will be much more effective than working through errors one at a time here.

**Comment by complexgenome** [comment]
@bentsherman  Thank you for your time and replies.
